src/charm.py

```diff
@@ -134,43 +147,50 @@ def _mysql(self) -> MySQL:
             self.k8s_helpers,
+            self,
         )
 
     @property
     def _pebble_layer(self) -> Layer:
         """Return a layer for the mysqld pebble service."""
-        return Layer(
-            {
-                "summary": "mysqld services layer",
-                "description": "pebble config layer for mysqld safe and exporter",
-                "services": {
-                    MYSQLD_SAFE_SERVICE: {
-                        "override": "replace",
-                        "summary": "mysqld safe",
-                        "command": MYSQLD_SAFE_SERVICE,
-                        "startup": "enabled",
-                        "user": MYSQL_SYSTEM_USER,
-                        "group": MYSQL_SYSTEM_GROUP,
-                        "kill-delay": "24h",
-                    },
-                    MYSQLD_EXPORTER_SERVICE: {
-                        "override": "replace",
-                        "summary": "mysqld exporter",
-                        "command": "/start-mysqld-exporter.sh",
-                        "startup": "enabled",
-                        "user": MYSQL_SYSTEM_USER,
-                        "group": MYSQL_SYSTEM_GROUP,
-                        "environment": {
-                            "DATA_SOURCE_NAME": (
-                                f"{MONITORING_USERNAME}:"
-                                f"{self.get_secret('app', MONITORING_PASSWORD_KEY)}"
-                                f"@unix({MYSQLD_SOCK_FILE})/"
-                            ),
-                        },
-                    },
+        layer = {
+            "summary": "mysqld services layer",
+            "description": "pebble config layer for mysqld safe and exporter",
+            "services": {
+                MYSQLD_SAFE_SERVICE: {
+                    "override": "replace",
+                    "summary": "mysqld safe",
+                    "command": MYSQLD_SAFE_SERVICE,
+                    "startup": "enabled",
+                    "user": MYSQL_SYSTEM_USER,
+                    "group": MYSQL_SYSTEM_GROUP,
+                    "kill-delay": "24h",
+                },
+                MYSQLD_EXPORTER_SERVICE: {
+                    "override": "replace",
+                    "command": "pwd",  # a simple command that does nothing
+                },
+            },
+        }
+
+        if self._has_cos_relation:
+            layer["services"][MYSQLD_EXPORTER_SERVICE] = {
+                "override": "replace",
+                "summary": "mysqld exporter",
+                "command": "/start-mysqld-exporter.sh",
+                "startup": "enabled",
+                "user": MYSQL_SYSTEM_USER,
+                "group": MYSQL_SYSTEM_GROUP,
+                "environment": {
+                    "DATA_SOURCE_NAME": (
+                        f"{MONITORING_USERNAME}:"
+                        f"{self.get_secret('app', MONITORING_PASSWORD_KEY)}"
+                        f"@unix({MYSQLD_SOCK_FILE})/"
+                    ),
```

Comment:
why not just make `startup` a function of cos_relation, i.e.:
```suggestion
                MYSQLD_EXPORTER_SERVICE: {
                    layer["services"][MYSQLD_EXPORTER_SERVICE] = {
                        "override": "replace",
                        "summary": "mysqld exporter",
                        "command": "/start-mysqld-exporter.sh",
                        "startup":  "enabled" if self._has_cos_relation else "disabled",
                        "user": MYSQL_SYSTEM_USER,
                        "group": MYSQL_SYSTEM_GROUP,
                        "environment": {
                            "DATA_SOURCE_NAME": (
                                f"{MONITORING_USERNAME}:"
                                f"{self.get_secret('app', MONITORING_PASSWORD_KEY)}"
                                f"@unix({MYSQLD_SOCK_FILE})/"
                            ),
```

Reply:
addressed in deecb90ea97e5fdeb7c0ec31d93943338540d81a. `startup` only controls if the service should be started when pebble starts. we also need to `container.stop()` to inactivate the service

Reply:
not entirely sure, but i saw replan() called but the `mysqld-exporter` service still enabled. https://ops.readthedocs.io/en/latest/#ops.Container.replan indicates possibly that it only restarts running services, and starts enabled services.

Reply:
that is indeed correct: `replan()` did not stop the service:

```
root@mysql-0:/# pebble plan
services:
    mysqld_exporter:
        summary: mysqld exporter
        startup: disabled
        override: replace
        command: /start-mysqld-exporter.sh
        environment:
            DATA_SOURCE_NAME: monitoring:jj7ojopaCYnnyHZgcNcF8sNc@unix(/var/run/mysqld/mysqld.sock)/
        user: mysql
        group: mysql
    mysqld_safe:
        summary: mysqld safe
        startup: enabled
        override: replace
        command: mysqld_safe
        user: mysql
        group: mysql
        kill-delay: 24h0m0s
    promtail:
        summary: promtail
        startup: disabled
        override: replace
        command: /opt/promtail/promtail-static-amd64 -config.file=/etc/promtail/promtail_config.yaml
root@mysql-0:/# pebble services
Service          Startup   Current  Since
mysqld_exporter  disabled  active   today at 18:48 UTC
mysqld_safe      enabled   active   today at 18:47 UTC
promtail         disabled  active   today at 18:48 UTC
```